File: src/main/java/net/ccbluex/liquidbounce/utils/render/IconUtils.java
/*
 * FDPClient Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge by LiquidBounce.
 * https://github.com/SkidderMC/FDPClient/
 *
 * This util sets a mac os dock icon
 * please include this header with all copies of this class
 * https://github.com/GlassMods/MinecraftSnippets/tree/main/MacOS-Dock-Icon-Java8
 */
package net.ccbluex.liquidbounce.utils.render;

import org.apache.commons.io.IOUtils;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.IOException;
import java.io.InputStream;
import java.nio.ByteBuffer;

public class IconUtils {

    public static ByteBuffer[] fav() {
        InputStream[] streams = new InputStream[]{
                IconUtils.class.getResourceAsStream("/assets/minecraft/fdpclient/16.png"),
                IconUtils.class.getResourceAsStream("/assets/minecraft/fdpclient/32.png"),
                IconUtils.class.getResourceAsStream("/assets/minecraft/fdpclient/64.png"),
                IconUtils.class.getResourceAsStream("/assets/minecraft/fdpclient/256.png")
        };
        ByteBuffer[] buffer = new ByteBuffer[streams.length];

        try {
            for (int i = 0; i < streams.length; i++) {
                buffer[i] = readImageToBuffer(streams[i]);
            }
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            for (InputStream s : streams) {
                IOUtils.closeQuietly(s);
            }
        }

        return buffer;
    }

    public static ByteBuffer readImageToBuffer(InputStream imageStream) throws IOException {
        BufferedImage var2 = ImageIO.read(imageStream);
        int[] var3 = var2.getRGB(0, 0, var2.getWidth(), var2.getHeight(), null, 0, var2.getWidth());
        ByteBuffer var4 = ByteBuffer.allocate(4 * var3.length);
        int[] var5 = var3;
        int var6 = var3.length;

        for (int var7 = 0; var7 < var6; ++var7) {
            int var8 = var5[var7];
            var4.putInt(var8 << 8 | var8 >> 24 & 255);
        }

        var4.flip();
        return var4;
    }

}

File: src/main/java/net/ccbluex/liquidbounce/utils/render/IconUtils.kt
/*
 * FDPClient Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/SkidderMC/FDPClient/
 */
package net.ccbluex.liquidbounce.utils.render

import net.ccbluex.liquidbounce.FDPClient.CLIENT_NAME
import net.ccbluex.liquidbounce.utils.client.ClientUtils
import net.ccbluex.liquidbounce.utils.io.flipSafely
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import org.lwjgl.opengl.Display
import java.io.IOException
import java.io.InputStream
import java.nio.ByteBuffer
import java.util.*
import javax.imageio.ImageIO

@SideOnly(Side.CLIENT)
object IconUtils {
    @JvmStatic
    fun initLwjglIcon(): Boolean {
        val icons = getFavicon()?.filterNotNull()?.toTypedArray()
        if (!icons.isNullOrEmpty()) {
            Display.setIcon(icons)
            return true
        }
        return false
    }

    fun getFavicon() =
        IconUtils::class.java.runCatching {
            arrayOf(
                readImageToBuffer(getResourceAsStream("/assets/minecraft/${CLIENT_NAME.lowercase()}/32.png")),
                readImageToBuffer(getResourceAsStream("/assets/minecraft/${CLIENT_NAME.lowercase()}/32.png")),
                readImageToBuffer(getResourceAsStream("/assets/minecraft/${CLIENT_NAME.lowercase()}/64.png"))
            )
        }.onFailure {
            ClientUtils.LOGGER.warn("Failed to load icons", it)
        }.getOrNull()

    @Throws(IOException::class)
    private fun readImageToBuffer(imageStream: InputStream?): ByteBuffer? {
        val bufferedImage = ImageIO.read(imageStream ?: return null)
        val rgb = bufferedImage.getRGB(0, 0, bufferedImage.width, bufferedImage.height, null, 0, bufferedImage.width)
        val byteBuffer = ByteBuffer.allocate(4 * rgb.size)
        for (i in rgb)
            byteBuffer.putInt(i shl 8 or (i shr 24 and 255))
        byteBuffer.flipSafely()
        return byteBuffer
    }
}